### ISSUE TITLE
A call to SSPI failed

### ISSUE THREAD

**Issue by justarandomuser** [Orange-Cyberdefense/GOAD#58]
I was consistently getting this error:

```
TASK [groups_domains : Add a domain user/group from another Domain in the multi-domain forest to a domain group : {{domain_server}}] **************************************************************************************************
changed: [192.168.56.11] => (item={'key': 'AcrossTheSea', 'value': ['essos.local\\daenerys.targaryen']})
FAILED - RETRYING: [192.168.56.10]: Add a domain user/group from another Domain in the multi-domain forest to a domain group : {{domain_server}} (3 retries left).
changed: [192.168.56.12] => (item={'key': 'DragonsFriends', 'value': ['sevenkingdoms.local\\tyron.lannister', 'essos.local\\daenerys.targaryen']})
ok: [192.168.56.12] => (item={'key': 'Spys', 'value': ['sevenkingdoms.local\\Small Council']})
FAILED - RETRYING: [192.168.56.10]: Add a domain user/group from another Domain in the multi-domain forest to a domain group : {{domain_server}} (2 retries left).
FAILED - RETRYING: [192.168.56.10]: Add a domain user/group from another Domain in the multi-domain forest to a domain group : {{domain_server}} (1 retries left).
An exception occurred during task execution. To see the full traceback, use -vvv. The error was:    at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.BeginProcessing()
failed: [192.168.56.10] (item={'key': 'AcrossTheNarrowSea', 'value': ['essos.local\\daenerys.targaryen']}) => {"ansible_loop_var": "item", "attempts": 3, "changed": false, "item": {"key": "AcrossTheNarrowSea", "value": ["essos.local\\daenerys.targaryen"]}, "msg": "Unhandled exception while executing module: A call to SSPI failed, see inner exception."}
```

An advanced google search led me to this URL:
[stackoverflow link](https://stackoverflow.com/questions/37925505/a-call-to-sspi-failed-see-inner-exception-the-local-security-authority-cannot)

I opted for the reg key mod and added the following between the "synchroizes all domains" and the "Add a domain user/group from another Domain...." tasks in the file "ansible/roles/groups_domains/tasks/main.yml":

```
- name: "Allow downgrade of TLS"
  ansible.windows.win_regedit:
    path: HKLM:\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\KeyExchangeAlgorithms\Diffie-Hellman
    name: ClientMinKeyBitLength
    data: 00000200
    type: dword
```

And now it's working. I can drop this in a pull request if desired...

**Comment by justarandomuser** [comment]
I think there might be more to it than that.
Running just the ad-relations.yml playbook now works but running the full main.yml one still fails in the same way.

**Comment by paularquette** [comment]
I tried this and it did not work for me, even just running the ad-relations.yml file

**Comment by justarandomuser** [comment]
Started getting this again. Not sure how to proceed.

**Comment by handbye** [comment]
same issue
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was:    at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.BeginProcessing()
failed: [dc01] (item={'key': 'AcrossTheNarrowSea', 'value': ['essos.local\\daenerys.targaryen']}) => {"ansible_loop_var": "item", "attempts": 3, "changed": false, "item": {"key": "AcrossTheNarrowSea", "value": ["essos.local\\daenerys.targaryen"]}, "msg": "Unhandled exception while executing module: A call to SSPI failed, see inner exception."}

```

**Comment by justarandomuser** [comment]
It may be completely unrelated but I think I had this issue reduce when I assigned a bit more (2gb) to each VM.
Could you try bumping DC01 up just a bit and see if it reproduces?

**Comment by verybiigcat** [comment]
Finally managed to solve this my end.
I was able to get the root cause of that error by running the playbook in verbose (-vvv) mode, which revealed that the "inner exception" there was "The clocks on the client and server machines are skewed".
Indeed one of the machine (DC03) had a different time set and wasn't synced with the other DCs.
I fixed that by running the following command on DC03:
w32tm /config /manualpeerlist:"kingslanding.sevenkingdoms.local" /syncfromflags:manual /reliable:yes /update
and then rerunning the task (see PS below).

PS: To be able to debug and rerun tasks I added "debugger: on_failed" to the task ("Add a domain user/group ...") in ~/GOAD/ansible/roles/groups_domains/tasks/main.yml
